include/RequestLimiter.inc.php

```diff
@@ -0,0 +1,168 @@
+<?php
+/*
+	This code is based on Stripe request limiter:
+	https://stripe.com/blog/rate-limiters
+*/
+
+// Notice: If we are using lua scripts, we shouldn't use any PHP serializer
+// for this redis connection, because inside lua scripts they can't be deserialized,
+// therefore we can get unexpected behaviour
+
+class Z_RequestLimiter {
+	// Lua script is from https://gist.github.com/ptarjan/e38f45f2dfe601419ca3af937fff574d
+	protected $request_rate_limiter_lua = '
+	local tokens_key = KEYS[1]
+	local timestamp_key = KEYS[2]
+	
+	local rate = tonumber(ARGV[1])
+	local capacity = tonumber(ARGV[2])
+	local now = tonumber(ARGV[3])
+	local requested = tonumber(ARGV[4])
+	
+	local fill_time = capacity/rate
+	local ttl = math.floor(fill_time*2)
+	
+	local last_tokens = tonumber(redis.call("get", tokens_key))
+	if last_tokens == nil then
+		last_tokens = capacity
+	end
+	
+	local last_refreshed = tonumber(redis.call("get", timestamp_key))
+	if last_refreshed == nil then
+		last_refreshed = 0
+	end
+	
+	local delta = math.max(0, now-last_refreshed)
+	local filled_tokens = math.min(capacity, last_tokens+(delta*rate))
+	local allowed = filled_tokens >= requested
+	local new_tokens = filled_tokens
+	if allowed then
+		new_tokens = filled_tokens - requested
+	end
+	
+	redis.call("setex", tokens_key, ttl, new_tokens)
+	redis.call("setex", timestamp_key, ttl, now)
+	
+	return { allowed, new_tokens }';
+	
+	// Lua script is from https://gist.github.com/ptarjan/e38f45f2dfe601419ca3af937fff574d
+	protected $concurrent_requests_limiter_lua = '
+	local key = KEYS[1]
+	
+	local capacity = tonumber(ARGV[1])
+	local timestamp = tonumber(ARGV[2])
+	local id = ARGV[3]
+	
+	local count = redis.call("zcard", key)
+	local allowed = count < capacity
+	
+	if allowed then
+		redis.call("zadd", key, timestamp, id)
+	end
+	
+	return { allowed, count }';
+	
+	public function __construct() {
+		$this->redis = Z_Redis::get();
```

Comment:
This will throw if `$REDIS_HOSTS` is undefined or empty in config.inc.php. We should just skip rate limiting entirely if there's no Redis server defined, and that may be true for most Redis-based features. So maybe this should return `false` instead of throwing, and it can be up to the callers?

(If `$REDIS_HOSTS` is undefined, we should log a warning in Z_Redis. If it's an empty array, I don't think we need to log anything, and calling functions can decide if not having Redis is a problem.)